File: proxion-messenger-core/src/proxion_messenger_core/device_cert.py
"""Account device certificates for multi-device linking (delegation model).

An account's canonical identity is its primary device's ``did:key``
(``account_did``). A secondary device generates its OWN Ed25519 keypair
(``device_did``) and is authorized to act for the account by a
``DeviceCertificate`` that the account's primary key *signs*. The gateway
verifies this certificate to admit the device as a session of the account.

This is delegation, NOT key-copying: the account's private key never leaves the
primary device (it only signs). That is a hard requirement here because device
identity keys are stored non-extractable by design (browser WebCrypto R9.1
hardening), so the account secret physically cannot be exported to a new device.

Canonical signing format
-------------------------
A length-prefixed concatenation of the four bound fields, identical byte-for-byte
to what the browser produces so a JS-issued cert verifies here:

    b"|".join( len(part).to_bytes(2,"big") + part for part in [
        account_did, device_did, str(issued_at), str(expires_at) ] )

Timestamps are integer seconds (no float formatting ambiguity across languages).
"""
from __future__ import annotations

import base64
import time
from typing import Optional

from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

from .didkey import did_to_pub_key, pub_key_to_did

# A device certificate may not outlive this window; the primary can re-issue.
MAX_TTL_DAYS = 400


def _canonical(account_did: str, device_did: str, issued_at: int, expires_at: int) -> bytes:
    parts = [
        account_did.encode(),
        device_did.encode(),
        str(int(issued_at)).encode(),
        str(int(expires_at)).encode(),
    ]
    return b"|".join(len(p).to_bytes(2, "big") + p for p in parts)


def issue_device_cert(
    account_priv: Ed25519PrivateKey,
    device_did: str,
    ttl_days: int = 365,
    now: Optional[float] = None,
) -> dict:
    """Sign a certificate authorizing ``device_did`` to act for the account.

    ``account_priv`` is the account's (primary device's) Ed25519 private key;
    the account_did is derived from its public key so the two can never drift.
    """
    if ttl_days <= 0 or ttl_days > MAX_TTL_DAYS:
        raise ValueError(f"ttl_days must be in 1..{MAX_TTL_DAYS}")
    if not device_did.startswith("did:key:"):
        raise ValueError("device_did must be a did:key")
    account_pub = account_priv.public_key().public_bytes_raw()
    account_did = pub_key_to_did(account_pub)
    issued = int(now if now is not None else time.time())
    expires = issued + ttl_days * 86400
    sig = account_priv.sign(_canonical(account_did, device_did, issued, expires))
    return {
        "@type": "ProxionDeviceCert",
        "version": 1,
        "account_did": account_did,
        "device_did": device_did,
        "issued_at": issued,
        "expires_at": expires,
        "signature": base64.b64encode(sig).decode("ascii"),
    }


def verify_device_cert(
    cert: dict,
    expected_device_did: Optional[str] = None,
    expected_account_did: Optional[str] = None,
    now: Optional[float] = None,
) -> Optional[str]:
    """Verify a device certificate.

    Returns the ``account_did`` the certificate authorizes for on success, or
    ``None`` on any failure (bad shape, forged/absent signature, expired, or a
    mismatch against an expected device/account binding). Never raises.

    ``expected_device_did`` should be the DID the connection just proved control
    of via the auth challenge — this is what stops one device replaying another
    device's certificate.
    """
    try:
        if not isinstance(cert, dict):
            return None
        account_did = cert.get("account_did", "")
        device_did = cert.get("device_did", "")
        issued_at = int(cert.get("issued_at", 0))
        expires_at = int(cert.get("expires_at", 0))
        sig_b64 = cert.get("signature", "")
        if not account_did or not device_did or not sig_b64:
            return None
        if not account_did.startswith("did:key:") or not device_did.startswith("did:key:"):
            return None
        if expected_device_did is not None and device_did != expected_device_did:
            return None
        if expected_account_did is not None and account_did != expected_account_did:
            return None
        now_i = int(now if now is not None else time.time())
        if expires_at <= issued_at or expires_at <= now_i:
            return None
        if expires_at - issued_at > MAX_TTL_DAYS * 86400:
            return None
        pub = Ed25519PublicKey.from_public_bytes(did_to_pub_key(account_did))
        pub.verify(
            base64.b64decode(sig_b64),
            _canonical(account_did, device_did, issued_at, expires_at),
        )
        return account_did
    except Exception:
        # Any failure (InvalidSignature, malformed base64/DID, bad shape) → reject.
        return None

File: web/device-cert.js
// device-cert.js — account device certificates for multi-device linking.
//
// Mirrors proxion_messenger_core/device_cert.py BYTE-FOR-BYTE so a certificate
// issued here (the primary device signs with its non-extractable clientDid key
// via WebCrypto) verifies on the Python gateway, and vice versa.
//
// Delegation, not key-copying: the account's private key only signs; it never
// leaves the primary device (device identity keys are non-extractable, R9.1).
//
// Canonical signing bytes (identical to the Python _canonical):
//   for each of [account_did, device_did, str(issued), str(expires)]:
//     2-byte big-endian length + utf8(part)
//   joined by a single 0x7C ("|") byte between the length-prefixed chunks.

const _ENC = new TextEncoder();
const MAX_TTL_DAYS = 400;
const _B58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

function _b58decode(str) {
    const map = {};
    for (let i = 0; i < _B58.length; i++) map[_B58[i]] = i;
    const bytes = [0];
    for (const ch of str) {
        const val = map[ch];
        if (val === undefined) throw new Error("bad base58 char");
        let carry = val;
        for (let j = 0; j < bytes.length; j++) {
            carry += bytes[j] * 58;
            bytes[j] = carry & 0xff;
            carry >>= 8;
        }
        while (carry > 0) { bytes.push(carry & 0xff); carry >>= 8; }
    }
    // leading '1's are leading zero bytes
    for (let k = 0; k < str.length && str[k] === "1"; k++) bytes.push(0);
    return new Uint8Array(bytes.reverse());
}

function _didToPubBytes(did) {
    if (!did.startsWith("did:key:z")) throw new Error("not a did:key");
    const mc = _b58decode(did.slice("did:key:z".length));
    if (mc.length < 34 || mc[0] !== 0xed || mc[1] !== 0x01) throw new Error("not an ed25519 did:key");
    return mc.slice(2, 34);
}

function _canonical(accountDid, deviceDid, issuedAt, expiresAt) {
    const parts = [accountDid, deviceDid, String(issuedAt), String(expiresAt)].map(s => _ENC.encode(s));
    const chunks = parts.map(p => {
        const c = new Uint8Array(2 + p.length);
        c[0] = (p.length >> 8) & 0xff;
        c[1] = p.length & 0xff;
        c.set(p, 2);
        return c;
    });
    const total = chunks.reduce((a, c) => a + c.length, 0) + (chunks.length - 1);
    const out = new Uint8Array(total);
    let off = 0;
    chunks.forEach((c, i) => {
        if (i > 0) out[off++] = 0x7c; // "|"
        out.set(c, off);
        off += c.length;
    });
    return out;
}

function _b64std(bytes) {
    let s = "";
    for (const b of bytes) s += String.fromCharCode(b);
    return btoa(s);
}

function _b64dec(s) {
    const bin = atob(s);
    const u = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; i++) u[i] = bin.charCodeAt(i);
    return u;
}

// Issue a certificate authorizing `deviceDid` to act for the account.
// `accountPrivKey` is a WebCrypto Ed25519 private CryptoKey (the primary's
// identity key); `accountDid` is its did:key (must correspond, or it won't
// verify). Returns a plain object ready to JSON-serialize.
export async function issueDeviceCert(accountPrivKey, accountDid, deviceDid, opts = {}) {
    const ttlDays = opts.ttlDays ?? 365;
    if (ttlDays <= 0 || ttlDays > MAX_TTL_DAYS) throw new Error(`ttlDays must be 1..${MAX_TTL_DAYS}`);
    if (!deviceDid.startsWith("did:key:")) throw new Error("device_did must be a did:key");
    const issued = Math.floor(opts.now == null ? Date.now() / 1000 : opts.now);
    const expires = issued + ttlDays * 86400;
    const canon = _canonical(accountDid, deviceDid, issued, expires);
    const sig = new Uint8Array(await crypto.subtle.sign("Ed25519", accountPrivKey, canon));
    return {
        "@type": "ProxionDeviceCert",
        version: 1,
        account_did: accountDid,
        device_did: deviceDid,
        issued_at: issued,
        expires_at: expires,
        signature: _b64std(sig),
    };
}

// Verify a certificate. Returns the authorized account_did on success, else null.
// Never throws. `expectedDeviceDid` should be this device's own DID so a cert
// meant for another device is not mistaken as ours.
export async function verifyDeviceCert(cert, opts = {}) {
    try {
        if (!cert || typeof cert !== "object") return null;
        const accountDid = cert.account_did || "";
        const deviceDid = cert.device_did || "";
        const signature = cert.signature || "";
        const issued = parseInt(cert.issued_at, 10);
        const expires = parseInt(cert.expires_at, 10);
        if (!accountDid || !deviceDid || !signature) return null;
        if (!accountDid.startsWith("did:key:") || !deviceDid.startsWith("did:key:")) return null;
        if (opts.expectedDeviceDid && deviceDid !== opts.expectedDeviceDid) return null;
        if (opts.expectedAccountDid && accountDid !== opts.expectedAccountDid) return null;
        if (!Number.isFinite(issued) || !Number.isFinite(expires)) return null;
        const nowI = Math.floor(opts.now == null ? Date.now() / 1000 : opts.now);
        if (expires <= issued || expires <= nowI) return null;
        if (expires - issued > MAX_TTL_DAYS * 86400) return null;
        const pub = await crypto.subtle.importKey(
            "raw", _didToPubBytes(accountDid), { name: "Ed25519" }, false, ["verify"]);
        const ok = await crypto.subtle.verify(
            "Ed25519", pub, _b64dec(signature), _canonical(accountDid, deviceDid, issued, expires));
        return ok ? accountDid : null;
    } catch {
        return null;
    }
}

// Exposed for tests / debugging — the exact bytes that get signed.
export const _internals = { _canonical, _didToPubBytes, _b58decode };
